compiler/rustc_mir_build/src/build/coverageinfo.rs

```diff
@@ -79,35 +89,91 @@ impl BranchInfoBuilder {
         }
     }
 
+    fn add_two_way_branch(
+        &mut self,
+        cfg: &mut CFG<'_>,
+        source_info: SourceInfo,
+        true_block: BasicBlock,
+        false_block: BasicBlock,
+    ) {
+        let true_marker = self.inject_block_marker(cfg, source_info, true_block);
+        let false_marker = self.inject_block_marker(cfg, source_info, false_block);
+
+        self.branch_arm_lists.push(vec![
+            BranchArm { span: source_info.span, marker: true_marker },
+            BranchArm { span: source_info.span, marker: false_marker },
+        ]);
+    }
+
+    pub(crate) fn add_match_arms(&mut self, cfg: &mut CFG<'_>, arms: Vec<MatchArm>) {
+        // Match expressions with 0-1 arms don't have any branches for their arms.
+        if arms.len() < 2 {
+            return;
+        }
+
+        // FIXME(#124118) The current implementation of branch coverage for
+        // match arms can't handle or-patterns or guards.
+        if arms.iter().any(|arm| arm.has_guard || arm.pre_binding_block.is_none()) {
+            return;
+        }
```

Comment:
This is where we currently avoid adding branch coverage for match expressions that contain any or-patterns or guards.

(This is an improvement over previous drafts, which would ICE on or-patterns, and give incorrect counts for match expressions containing guards.)

For or-patterns, we can probably dig deeper into the pattern and find a suitable block, but I'll have to pay close attention to make sure that what I'm doing is correct.

For guards, we need to track enough information to let the instrumentor subtract the guard-was-false path from the sum of subsequent arms.